File: app/src/main/java/ru/melod1n/vk/concurrent/TryCallback.java
package ru.melod1n.vk.concurrent;

import ru.melod1n.vk.common.AppGlobal;

public abstract class TryCallback implements Runnable {

    public abstract void ready() throws Exception;

    public abstract void done();

    public abstract void error(Exception e);

    @Override
    public void run() {
        try {
            ready();
        } catch (Exception e) {
            e.printStackTrace();

            AppGlobal.handler.post(() -> error(e));
            return;
        }

        AppGlobal.handler.post(this::done);
    }
}

File: app/src/main/java/ru/melod1n/vk/concurrent/TryCallback.kt
package ru.melod1n.vk.concurrent

import ru.melod1n.vk.common.AppGlobal

abstract class TryCallback : Runnable {
    @Throws(Exception::class)
    abstract fun ready()

    abstract fun done()
    abstract fun error(e: Exception?)
    override fun run() {
        try {
            ready()
        } catch (e: Exception) {
            e.printStackTrace()
            AppGlobal.handler.post { error(e) }
            return
        }
        AppGlobal.handler.post { done() }
    }
}